### PR TITLE
REF: move convert_scalar out of cython

### PR DESCRIPTION
**Pull request by jbrockmendel**
There's nothing about this that particularly benefits from being in cython (I think until recently this was used within index.pyx) and its clearer in python.  Plus we get a slightly smaller/faster build.

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -3204,3 +3205,34 @@ def _putmask_preserve(nv, n):
         v = v.astype(dtype)
 
     return _putmask_preserve(v, n)
+
+
+def _convert_scalar_for_putitemlike(scalar, dtype: np.dtype):
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: maybe in pandas.core.dtypes.cast ?

**Comment by jreback** [comment]
needs a rebase as well

File: pandas/core/internals/blocks.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py]: Sure.  With convert_scalar out of libindex, we can move validate_numeric_casting to core.dtypes.cast as well (and after #31662 we'll be able to get rid of libindex.get_value_at, plus util.get_value_at and util.validate_indexer!)

Comment (jreback) [reply pandas/core/internals/blocks.py]: sgtm.

**Comment by jreback** [comment]
ping on green.

**Comment by jbrockmendel** [comment]
ping

**Pull request merged**